### ISSUE TITLE
Limit number of parallel compactions of different CFs

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#1295]
I I understand correctly, starting in commit 5aeeb0b3e8859a9687bc6fc3b41d4d46fd4ac93f, parallel compaction of different CFs is no longer limited: If you have 10 different CFs, you might have 10 compactions active in parallel (even more if we decide to do parallel compactions on on CF).

The downside of parallel compaction is that it requires quite a bit of memory, such as buffers for reading all the sstables, and when mutations are very large (before we support wide partitions properly), holding multiple huge mutations in memory.

In some cases, parallel compaction is important (e.g., to not block small compactions while a very large compaction is in progress), but I don't understand why it might be important to do an large number of compactions of same size CFs in parallel. As an example, consider two CFs that want to start a large compaction which would each take (separately) 10 minutes. The current code may start both together, and they will run 20 minutes. I'm suggesting it's better to start just one, have it finish in 10 minutes, and then start the second, that will finish after an additional 10 minutes.
Just don't forget that even with my suggestion, while a long compaction is running on any CF, we should allow a short compaction to start (on any CF).

**Comment by avikivity** [comment]
To summarize, compactions should be serialized if they are the same size class, and parallelized if not.  Is that a good description?

**Comment by nyh** [comment]
I think so...
There are two additional snags:
1. If different CFs are on different disks, it may be beneficial to compact them in parallel. But this is not an issue in our typical installation (where the multiple disks are joined into RAID).
2. We need some sort of fairness on the queuing of compactions that get deferred - so that one CF's constant compactions doesn't starve other CFs. It might not be enough to simply not do a compaction and have it retried again some undetermined amount of time later.

**Comment by slivne** [comment]
" and when mutations are very large (before we support wide partitions
properly), "
I think we can disregard that - we are already working toward a solution.

If we allow to use the mid result of a compaction (partial sstable being
written) while it is being run then it may be better to run them both in
parallel (I think cassandra does that) - instead of trying to order them
(what is the correct order ?) - the compaction may allow better read
performance.

**Comment by nyh** [comment]
@slivne you may be right about the being able to use partial results of compaction (though we don't have such code). But I'm not completely sure if there's a real advantage to finishing half the compaction of two CFs than completely finishing just one. I think in both cases you'll make some reads faster and some not - just in the serial compaction, this will bias will be easier to name (reads to one CF will be slower than to the other).

By the way, another downside of doing large compactions in parallel is that it needs more temporary disk space (in general, we can't delete the old sstables until we completed the compaction, so we have two copies of the data being compacted).

**Comment by avikivity** [comment]
With multi-tenancy you have to issue the compactions in parallel, because different tenants have different resource allocations.

For a single tenant serializing compaction (for a given size class) improves overall efficiency.  We should probably pick the compaction with the highest number of sstables; or maybe FIFO order is sufficient.

**Comment by slivne** [comment]
I think its also coupled to to compaction strategy and if you have
incremental repair or not

for example in leveled:
- you should probably compact l0 - as the data is spread through multiple
sstables and you are hitting a penalty of reading multiple files instead of
using level advantage.
- the difference between l1,l2,l3 is less clear to me.

**Comment by nyh** [comment]
@avikivity regarding multi-tenancy, if we do not overcommit resources at all, and promise each tenant a specific slice of the CPU and memory, then you are right.

But if we do overcommit, and bill each tenant for the resources they got, then I think we'll get better efficiency by serializing the compaction - even if it means one tenant's compaction gets delayed while another tenant's compaction is being done. One way to think about it: Imagine that we can bill a client for the amount of memory he actually uses - e.g., 1 cent per gigabyte*hour (some providers call this unit "GB Hour"). Imagine that while a certain compaction is in progress, it uses 100 MB. If this tenant was allowed to run this compaction to completion, it would finish in (say) 30 seconds. But if it is done in parallel with 4 other tenants, it will complete in 4 times longer. So the tenant will need to pay to hold that 100 MB in memory for 4 times longer. He would probably not appreciate it, and prefer that compactions of different tenants are staggered, not done concurrently.

**Comment by raphaelsc** [comment]
'For a single tenant serializing compaction (for a given size class) improves overall efficiency. We should probably pick the compaction with the highest number of sstables; or maybe FIFO order is sufficient.'

If that's what we want, we may go back to the approach of continuously running fiber. Now one fiber for each size class, and each size class would have its own queue (FIFO order). So we may have N parallel compaction per shard, where N is the number if size classes. However, Nadav apparently wants a better mechanism to guarantee fairness among column families.
